src/box/txn.c

```diff
@@ -850,10 +851,26 @@ txn_journal_entry_new(struct txn *txn)
 		if (stmt->row == NULL)
 			continue;
 
+		/*
+		 * Look at the flags stmt->row->flags. If the transaction
+		 * is synchronous, then set is_sync = true. This should
+		 * only be done on replicas. The master sets these flags
+		 * and independently decides whether the transaction is
+		 * synchronous or not.
+		 */
+
+		if (stmt->row->flags == (IPROTO_FLAG_COMMIT |
+		    IPROTO_FLAG_WAIT_SYNC | IPROTO_FLAG_WAIT_ACK)) {
+			is_sync = true;
+		}
+
 		if (stmt->row->type != IPROTO_NOP) {
 			is_fully_nop = false;
 			is_sync = is_sync || (stmt->space != NULL &&
-					      space_is_sync(stmt->space));
+					      space_is_sync(stmt->space)) ||
+					      txn->is_sync;
+
+			txn->is_sync = false;
```

Comment:
`txn->is_sync` needs to be cleared as `box_txn_set_sync()` sets `txn->is_sync = true`. It is done this way because `box.commit({is_sync = false})` cannot change the transaction. Since if the transaction was opened as synchronous (or there were synchronous spaces), it should remain so. Yes, indeed, it is not necessary to clear `txn->is_sync` on every row.

Reply:
> txn->is_sync needs to be cleared as box_txn_set_sync() sets txn->is_sync = true. Yes, indeed, it is not necessary to clear txn->is_sync after every row.

Why can't you leave it `true`?